spec/consensus/gossip/crdt.md

```diff
@@ -130,48 +129,178 @@ Consider the following definition for **Eventual Convergence**.
+- once an entry is added to the local view, deliver it.
 
 However, if entries can be removed, then the Tuple Space is actually weaker, since some entries may never be seen by some nodes, and should be easier to implement.
 We argue later that it can be implemented using Anti-Entropy or Epidemic protocols/Gossiping (not equal to the Gossip Communication property).
 We pointed out [previously](#on-the-need-for-the-gossip-communication-property) that the Gossip Communication property is overkill for Tendermint because it requires even stale messages to be delivered.
-Removing tuple is exactly how stale messages get removed.
+We remove entries corresponding to stale messages and never deliver them them.
 
 [^proof]:  TODO: do we need to extend here?
 
 ### When to remove
 
 > **TODO** Define conditions for tuple removal. Reference
 
+## The tuple space as a CRDT
+
+Conflict-free Replicated Data Types (CRDT) are distributed data structures that explore commutativity in update operations to achieve [Strong Eventual Consistency](https://en.wikipedia.org/wiki/Eventual_consistency#Strong_eventual_consistency).
+As an example of CRDT, consider a counter updated by increment operations, known as Grown only counter (G-Counter): as long as the same set of operations are executed by two replicas, their views of the counter will be the same, irrespective of the execution order.
+
+More relevant CRDT are the Grow only Set (G-Set), in which operations add elements to a set, and the [2-Phase Set](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#2P-Set_(Two-Phase_Set)) (2P-Set), which combines two G-Set to collect inclusions and exclusions to a set.
+
+CRDT may be defined in two ways, operation- and state-based.
+Operation-based CRDT use reliable communication to ensure that all updates (operations) are delivered to all replicas.
+If the reliable communication primitive precludes duplication, then applying all operations will lead to the same state, irrespective of the delivery order since operations are commutative.
+If duplications are allowed, then the operations must be made idempotent somehow.
+
+State-based CRDT do not rely on reliable communication.
+Instead they assume that replicas will compare their states and converge two-by-two using a merge function;
+as long as the function is commutative, associative and idempotent, the states will converge.
+For example, in the G-Set case, the merge operator is simply the union of the sets.
+
+The two approaches for converging the message sets in the Tendermint algorithm described [earlier](#nodes-state-as-a-tuple-space), without the deletion of entries, correspond to the operation- and state-based [Grow-only Set](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#G-Set_(Grow-only_Set)) CRDT;
+if removals must be handled, then using a 2P-Set is an option.
+To the best of our knowledge, no existing CRDT supports superseding of elements, until now.
```

Comment:
I would give some intuition about what does mean _supeseeding_ here. 

I believe that, on the contrary, this is the very essence of CRDTs, that is once a state $u$ is reached (whatever how), any of the prefixes of $u$ in the lattice are included/superseded.